[Manusha Galappaththi] iP

manu2002g opened the pull request:
# Prince

> Due tomorrow, do tomorrow - Unknown

DukePro frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

To get started using price right now
1. download it from [here](https://github.com/manu2002g/ip/releases/tag/A-Junit)
2. run the file from your computer
3. add your tasks
4. watch your life be **10x** easier 💰 

Features:
- [x] Add tasks
- [x] Search task list
- [ ] Edit tasks
- [ ] Major coding standard violations

Here's some sample code from the `Duke` class:
```java
public static void main(String[] args) {
        new Duke("data/tasks.txt").run();
    }
```

Commit:
Add Gradle support

manu2002g committed:
completed Level 1

manu2002g committed:
completed Level 2

manu2002g committed:
completed Level 3

manu2002g committed:
completed Level 4

manu2002g committed:
automated text ui testing

manu2002g committed:
completed Level 5

manu2002g committed:
completed Level 6

manu2002g committed:
cleaned up code + added exceptions

manu2002g committed:
updated test script

manu2002g committed:
Add local storage feature

manu2002g committed:
Add java.time.LocalDate

manu2002g committed:
Clean up code

manu2002g committed:
Update .gitignore

manu2002g committed:
Merge branch 'branch-Level-7'

manu2002g committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

manu2002g committed:
Fix date conversion bug in Parser

whtan99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+
 public class Duke {
+
+    static ArrayList<Task> list = new ArrayList<Task>();
```
```suggestion
    static ArrayList<Task> taskList = new ArrayList<Task>();
```

whtan99 replied:
Would adding task will help to make the code more readable to users?

waffledood replied:
Agree with @goalfix's comment here, `taskList` is commonly used by others as well :)

whtan99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+    static ArrayList<Task> list = new ArrayList<Task>();
+
+    public static void handle(String input) throws DukeException{
+        String[] split = input.split(" ");
```
Would terming `split` as `UserCommands` or `splitByDate` be clearer for the user?

whtan99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+                System.out.println("Congrats there's nothing on your list!");
+            } else {
+                System.out.println("Here's everything on your list rn:");
+                for (Integer i = 1; i <= list.size(); i++) {
+                    System.out.println(i.toString() + " " + list.get(i - 1));
+                }
+            }
+        } else if (command.equals("mark") || command.equals("unmark")) {
+            int idx = Integer.parseInt(split[1]);
+            if (idx > list.size()) {
+                throw new DukeException("The task number you've given is invalid");
+            } else {
+                Task t = list.get(idx - 1);
+                if (command.equals("unmark")) {
+                    t.makeNotDone();
+                    System.out.printf("Ok boss I've marked task %s as incomplete\n", split[1]);
+                } else {
+                    t.makeDone();
+                    System.out.printf("Woohoo! I've marked task %s as done\n", split[1]);
+                }
+                System.out.println(t.toString());
+            }
+        } else if (command.equals("delete")) {
+            if (list.size() == 0) {
+                throw new DukeException("There aren't any tasks left to delete");
+            } else {
+                int idx = Integer.parseInt(split[1]);
+                if (idx > list.size()) {
+                    throw new DukeException("The task number you've given is invalid");
+                } else {
+                    Task t = list.get(idx - 1);
+                    list.remove(idx - 1);
+                    System.out.printf("Alrighty I've removed this task\n");
+                    System.out.println(t.toString());
+                    System.out.printf("K you've now got %d tasks in the list\n", list.size());
+                }
+            }
+        } else {
+            throw new DukeException("I'm not sure what that means");
+        }
```
Would using a Case/Switch block be better for readability?

whtan99 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,76 @@
+
+    String monthToInt(String month) {
+        String r = "";
+        switch (month) {
+        case "Jan":
+            r = "01";
+            break;
+        case "Feb":
+            r = "02";
+            break;
+        case "Mar":
+            r = "03";
+            break;
+        case "Apr":
+            r = "04";
+            break;
+        case "May":
+            r = "05";
+            break;
+        case "Jun":
+            r = "06";
+            break;
+        case "Jul":
+            r = "07";
+            break;
+        case "Aug":
+            r = "08";
+            break;
+        case "Sep":
+            r = "09";
+            break;
+        case "Oct":
+            r = "10";
+            break;
+        case "Nov":
+            r = "11";
+            break;
+        case "Dec":
+            r = "12";
+            break;
```
I really like this case/switch statement! Makes it really easy for me to read 👍

whtan99 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,26 @@
+        return (isDone ? "X" : " "); // mark done task with X
+    }
+
+    public void  makeDone() {
```
Is there an extra space between `void` and `makeDone()`?

whtan99 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+            if (split.length == 1) {
+                throw new DukeException("It looks like you're missing the task description");
+            } else {
+                String[] arr = input.substring(6).split("/at ");
```
Would adding a `name/variable` to `input.substring(6)` be better for code readability?

whtan99 reviewed (commented):
Overall, looks good to me! 👍 
Some minor changes to variable names, as well as coding standards suggested

waffledood commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,76 @@
+        switch (month) {
+        case "Jan":
+            r = "01";
+            break;
+        case "Feb":
+            r = "02";
+            break;
+        case "Mar":
+            r = "03";
+            break;
+        case "Apr":
+            r = "04";
+            break;
+        case "May":
+            r = "05";
+            break;
+        case "Jun":
+            r = "06";
+            break;
+        case "Jul":
+            r = "07";
+            break;
+        case "Aug":
+            r = "08";
+            break;
+        case "Sep":
+            r = "09";
+            break;
+        case "Oct":
+            r = "10";
+            break;
+        case "Nov":
+            r = "11";
+            break;
+        case "Dec":
+            r = "12";
+            break;
+
+        }
+        return r;
```
Instead of having a very verbose block of code, you might want to consider using available classes & methods that can do the parsing of calendar months as integers. 

Classes such as `java.time.Month` & `java.util.Calendar` can do this. 

Refer to this [StackOverflow post](https://stackoverflow.com/questions/24355504/get-month-number-from-month-name) for more details.

waffledood commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+import java.util.Scanner;
+import java.util.ArrayList;
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.FileNotFoundException;
+import java.time.LocalDate;
```
Consider spacing out your imports, as per [guidelines](https://se-education.org/guides/conventions/java/intermediate.html#package-and-import-statements) 

```suggestion
import java.util.Scanner;
import java.util.ArrayList;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.FileNotFoundException;

import java.time.LocalDate;
```

waffledood reviewed (commented):
LGTM overall! 👍 
Just a few changes suggested, with regards to code quality & coding standard!

seanlaiys commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,18 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Deadline extends Task{
```
Would adding a white space after "Task" be better?

seanlaiys commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,13 @@
+public class Event extends Task{
```
Would adding a white space after "Task" be better?

riakhaitan replied:
Agreed with @seanlaiys, adding a white space after "Task" would be following coding standard.

seanlaiys commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,12 @@
+public class Todo extends Task{
```
Would adding a white space after the bracket be better?

seanlaiys reviewed (commented):
Overall great coding standard! Just some minor things to address.

riakhaitan commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,76 @@
+
+    Task parse() {
+        Task t = new Task("null");
+        switch (string.substring(0,3)) {
```
Good job following the coding standards while using switch

riakhaitan commented on src/main/java/DukeException.java:
```diff
@@ -0,0 +1,6 @@
+public class DukeException extends Exception {
```
Consider renaming to PrinceException since your chatbot is called Prince

riakhaitan commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,166 @@
+import java.io.FileNotFoundException;
+import java.time.LocalDate;
+
 public class Duke {
```
Change name to Prince to avoid confusion

riakhaitan reviewed (commented):
Coding standard followed

manu2002g committed:
Do More-OOP

manu2002g committed:
Add packages

manu2002g committed:
Add JUnit tests

manu2002g committed:
Add JavaDoc comments

manu2002g committed:
Clean up code

manu2002g committed:
Add find feature

manu2002g committed:
Complete JavaDocs

manu2002g committed:
Fix coding standard

manu2002g committed:
Complete find feature

manu2002g committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

manu2002g committed:
Merge branch 'branch-A-CodingStandard' into branch-Level-9

manu2002g committed:
Merge branch 'branch-A-CodingStandard'

manu2002g committed:
Merge branch 'add-gradle-support' of https://github.com/manu2002g/ip into branch-A-Gradle

manu2002g committed:
Add GUI

manu2002g committed:
Remove unnecessary imports

manu2002g committed:
Add assertions to the handle method

There may be bugs hidden in the handle method as it is one of the more
complicated methods.

Adding assertions helps to document assumptions that should hold at
that specific point in the code.

Let's add an assertion to make sure out is not an empty string.

manu2002g committed:
Improve code quality

The code that handled the creation of a Deadline was confusing and does
not follow SLAP. There are unnecessary imports and unclear variable
names.

Improving the code quality will make it more
readable and less confusing to debug.

Let's
* remove unnecessary imports
* make variable names more descriptive
* add comments explaining the logic of the code

manu2002g committed:
Merge pull request #2 from manu2002g/branch-A-CodeQuality

Improve code quality

manu2002g committed:
Merge branch 'master' of https://github.com/manu2002g/ip

manu2002g committed:
Merge branch 'branch-A-Assertions'

* branch-A-Assertions:
  Add assertions to the handle method

# Conflicts:
#	src/main/java/duke/Duke.java

manu2002g committed:
Add help command (C-help)

There is no documentation for new users to refer to and familiarise
themselves with Prince's features.

A help command gives quick access to a manual on how to use Prince.

Let's add a help command that lists out all possible commands, along
with a description and an example use

manu2002g committed:
Change GUI to make more effective use of space by removing
profile picture for the user

manu2002g committed:
Update bye command to save and add documentation

manu2002g committed:
Update readme

manu2002g committed:
Set theme jekyll-theme-minimal

manu2002g committed:
Update readme

manu2002g committed:
Update README

manu2002g committed:
Merge branch 'master' of https://github.com/manu2002g/ip

* 'master' of https://github.com/manu2002g/ip:
  Set theme jekyll-theme-minimal

manu2002g committed:
Change class structure to allow building fat jar file

manu2002g committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Complete find feature

# Conflicts:
#	src/main/java/duke/TaskList.java

manu2002g committed:
Improve Code Quality

Let's
* Update/shorten methods to use SLAP
* Remove unnecessary methods
